docs/secure_coding_and_deployment_hardening/atom_exhaustion.md

```diff
@@ -8,6 +8,15 @@ next:
   title: Serialisation and deserialisation
 ---
 
+* When converting strings or binaries to atoms, it’s crucial to ensure only a predefined set of atoms is
```

Comment:
I’m not fond of the word `crucial` here, as there may be good reasons to allow `to_existing_atom`. I think that this doesn't quite fit with the rest of the document where it sits and instead probably belongs as an extension of Background ¶3 (line 35,44). My suggestion would be to change that paragraph to:

```markdown
The best way to prevent atom exhaustion is by ensuring no new atom values are created at runtime: as long as any atom value required by the application is referenced in code, that value will be defined in the atom table when the code is loaded (e.g. at [release](releases) startup). The conversion of other types into atoms can then be constrained to only allow existing values using the `...to_existing_atom/1` function variants. Even safer is to use a lookup table that maps known values to atoms, as it not only safeguards against atom table exhaustion but ensures strict control over which atoms may be used in a given application context.
```

Reply:
Thanks for commenting! I'm sure we can find some middle ground between the original "[if the] set of desired atoms is known, consider converting them yourself" and the new "crucial." I'm not too hung up on the wording :-)

I'd prefer it to be at the top of the page though -- similar to other documents, putting important information applicable to both Elixir and Erlang first. Shoving this in the background risks having people use `to_existing_atom` when a lookup table would be a much better solution.

Reply:
@halostatue I agree that crucial might be the wrong word. I however also don't think that we need to go as much into detail as you suggested.

What do you both think about a softened version?

```markdown
* When converting strings or binaries to atoms, it’s often advisable to ensure only
  a predefined set of atoms is used. While functions like
  [`list_to_existing_atom/1`](https://erlang.org/doc/man/erlang.html#list_to_existing_atom-1) and
  [`String.to_existing_atom/1`](https://hexdocs.pm/elixir/String.html#to_existing_atom/1)
  can prevent the creation of new atoms, using a lookup table to map known values
  to atoms offers an additional layer of safety. This not only helps prevent atom table
  exhaustion but also ensures stricter control over which atoms are allowed, minimizing
  the risk of introducing unexpected or unwanted values into your application.
```

Reply:
If it's at the top of the page, my suggestion would be to make it not a single list entry, but its own paragraph with no list marker. This may or may not match other documents (it's been a while since I’ve looked). I’d *personally* write it with a more prescriptive tone:

```markdown
It's recommended to restrict the conversion of strings or binaries to atoms to a predefined set of atoms. This is best done using lookup tables to map known values to atoms as it prevents atom table exhaustion and ensures stricter control over the atoms allowed in your application. Functions like [list_to_existing_atom/1](…) and [String.to_existing_atom/1](…) may be used to ensure that no new atoms will be created in other cases where lookup tables are inadvisable.
```

I’m not *really* happy with that wording, but that's really because this is a special case of "handling untrusted external input", and whenever there should be a restricted input, it should almost always be handled with some form of lookup table.

Reply:
@halostatue I like your proposed paragraph besides the last word “inadvisable”. Maybe impractical or something like that.

It’s not that we don’t advise the functions at all. It might just not fit your use case.

Reply:
I just reformatted the paragraph for easier reading. Yes, I think that "impractical" is a better choice. It *might* be worth noting specifically that this is a special case of "handling (potentially) untrusted external input" and that's why lookup tables are the best choice, but my wording brain is…not at full speed right now.

Reply:
+1 for "impractical." I'm not sure we need to mention it as a special case of handling untrusted external input, as it feels a bit redundant given the page's nature, and either way it's a good idea to indicate what's expected even for trusted inputs. I'm not against it though :-)

Reply:
Are we all in agreement now? Then please update to the latest version.

The next working group call will be on Wednesday. I would like to tell people to add their feedback until then or we'll go ahead and merge.

Reply:
Yeah, I think so. I've updated the PR :-)